gendarme/man/gd2i.1.in
.\" 
.\" gd2i manual page.
.\"
.\" Authors:
.\"	Sebastien Pouliot  <sebastien@ximian.com>
.\"
.\" Copyright (C) 2010 Novell, Inc (http://www.novell.com)
.\"

.de Sp
.if t .sp .5v
.if n .sp
..
.TH gd2i "Gendarme @GENDARME_VERSION@"
.SH NAME
Gendarme Defect To Ignore List
.SH SYNOPSIS
.PP
.B gd2i defects.xml list.ignore [--extra-check] [--syntax-check] [--quiet] [--version] [--help]
.SH DESCRIPTION
\fIgd2i\fP is small tool to help you create, or update, an ignore-list 
baseline from an XML list of defects (generated by \fIGendarme\fP).
.SH OPTIONS
.TP
.I "defects.xml"
The, XML formatted, list of defects produced by Gendarme on your project.
.TP
.I "list.ignore"
The, text, file listing ignored defects entries for your project. If the 
file does not already exists (i.e. an update) then it will be created.
See gendarme(5) for the format of this file.
.TP
.I "--extra-check"
Report ignore entries, from an existing 'list.ignore' file, that does 
not match any defect, from the 'defects.xml' file.
.TP
.I "--syntax-check"
Report syntax error found when processing an existing 'list.ignore' file.
.TP
.I "--quiet"
Used to disable progress and other information which is normally written to stdout.
.TP
.I "--version"
Display the tool's version number. This will match the Mono version number
that this copy of Gendarme was released with.
.TP
.I "--help"
Show help about the command-line options.
.SH EXIT STATUS
.TP
.I 0
The tool returns 0 when no error has occurred. 
.TP
.I 1
The tool returns 1 if some command-line parameter were not valid.
.TP
.I 2
The tool execution was interrupted by a non-handled exception. This
is likely a bug inside the tool and should be reported on Novell's
bugzilla (http://bugzilla.novell.com) or on the mailing-list.
.SH COPYRIGHT
Copyright (C) 2010 Novell, Inc (http://www.novell.com)
.SH MAILING LISTS
Mailing lists are listed at the
http://groups.google.com/group/gendarme
.SH WEB SITE
http://www.mono-project.com/Gendarme
.SH SEE ALSO
.BR gendarme(1), gendarme(5), mono(1)


ChangeLog
2010-06-07  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add new gendarme directories

2010-04-18  Sebastien Pouliot  <sebastien@ximian.com>

	* README: Add instructions for building from SVN

2010-03-02  Marek Habersack  <mhabersack@novell.com>

	* configure.in: added check for dmcs, defines DMCS_PRESENT
	conditional if found. Needed to build docbrowser with dmcs now
	that mono from trunk tools are built with dmcs.

2010-02-23  Michael Hutchinson  <mhutchinson@novell.com>

	* configure.in: Add docbrowser/theme-icons/Makefile.am

2010-02-03  Michael Hutchinson  <mhutchinson@novell.com>

	* Makefile.am:
	* configure.in: Capture ACLOCAL_FLAGS for autoreconf
	  so building is less painful on Mac.

2010-02-03  Michael Hutchinson  <mhutchinson@novell.com>

	* configure.in:
	* asn1view/gtk/Makefile.am
	* docbrowser/BrowserWidget.cs:
	* docbrowser/Makefile.am: Make Gnome# optional.

	* autogen.sh: Pass ACLOCAL_FlAGS to aclocal.

2009-11-29  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Now 2.7

2009-09-24 Gonzalo Paniagua Javier <gonzalo@novell.com>

	* configure.in: avoid a lot of warnings.

2009-09-20  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add gendarme.5.in

2008-12-15  Miguel de Icaza  <miguel@novell.com>

	* configure.in: Detect the proper LIB_PREFIX and LIB_SUFIX to fix
	the crasher for gsharp when trying to Plot on other systems.

2009-06-05  Michael Hutchinson  <mhutchinson@novell.com>

	* configure.in: Fix build.

2009-06-03  Jonathan Pryor <jpryor@novell.com>

	* Makefile.am: Add webdoc to the build.

2008-12-02  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Remove AC_OUTPUT on gnunit files since they break
	make dist

2008-12-01  Atsushi Enomoto  <atsushi@ximian.com>

	* Makefile.am : remove gnunit. It does not build anymore.

2008-11-17  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Produce gendarme.exe.config from .in file (modified
	patch from Jesse Jones).

2008-11-10  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Bump to 2.3

2008-11-09  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add Gendarme framework unit test makefile generation.

2008-10-10  Massimiliano Mantione  <massi@ximian.com>
	* Mono.Profiler/Mono.Profiler.mds, Mono.Profiler/mprof-runtime/:
	Preliminary (but working) code to control the profiler at runtime.

2008-10-09 Daniel Nauck <dna@mono-project.de>

        * Makefile.am, configure.in:
		add gsharp to the build

2008-08-26  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add Gendarme's Design.Generic rules makefile support.

2008-08-25  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add Gendarme's Security.Cas rules makefile support.

2008-08-16  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add Gendarme's Design.Linq rules makefile support.

2008-08-12  Massimiliano Mantione  <massi@ximian.com>
	* Mono.Profiler/Makefile.am, Mono.Profiler/adjust-makefiles.sh:
	More fixes for the Mono.Profiler integration.

2008-08-12  Massimiliano Mantione  <massi@ximian.com>
	* Makefile.am, configure.in: Integrated Mono.Profiler in the build.
	* Mono.Profiler/configure.ac, Mono.Profiler/autogen.sh:
	Deleted (no longer needed).
	* Mono.Profiler/adjust-makefiles.sh: Added this script to fix the
	makefiles again if they are regenerated from Monodevelop.

2008-08-11  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Bump Gendarme version to match the Mono release.

2008-08-11  Massimiliano Mantione  <massi@ximian.com>

	* Makefile.am: reverted previous commit (it breaks the build).

2008-08-11  Massimiliano Mantione  <massi@ximian.com>

	* Makefile.am: Added Mono.Profiler subdirectory.

2008-07-14  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Bump Gendarme to 0.2.0.0 for the Mono 2.0 branch.

2008-06-14  Andreia Gaita <avidigal@novell.com> 

	* configure.in: Enable webkit and mono.webbrowser detection.

2008-04-27  Nestor Salceda  <nestor.salceda@gmail.com>

	* configure.in: Add Maintainability setup to the build.

2008-03-22  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add Gendarme's Wizard and Win32 setup to the build

2008-03-12  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Add Gendarme's Test.Rules makefile support.

2008-03-08  Andreia Gaita  <avidigal@novell.com>

	* configure.in (docbrowser): Add mono-webbrowser check, and tweak
	the detected browser engine message to list what is available.

2008-03-07  Alp Toker  <alp@nuanti.com>

	* configure.in: Add no-op webkit-sharp check (for docbrowser).

2008-02-16  Sebastien Pouliot  <sebastien@ximian.com>

	* configure.in: Gendarme does not need the .xml[.in] files anymore :)

2008-02-01  Andrew Jorgensen  <ajorgensen@novell.com>

	* configure.in: get the path to Mono.Cecil.dll during configure

2008-01-28  Wade Berrier  <wberrier@novell.com>

	* configure.in: version bump -> 1.9

2007-12-12  Mike Kestner  <mkestner@novell.com>

	* configure.in: update pkgconfig checks for gtkhtml-sharp-3.14
	and for gconf without Requires: gnome-sharp.
	* docbrowser/Makefile.am: update for GACd gtkhtml-sharp-3.14.

2007-11-09  Wade Berrier  <wberrier@novell.com>

	* configure.in: define MONO_TOOSL in a mcs friendly way (svn revs)
	version bump -> 1.2.6

2007-11-08  Mike Kestner  <mkestner@novell.com>

	* configure.in: configure checks for gtkhtml-sharp-3.14
	with fallback to 2.0.

2007-04-19  Wade Berrier  <wberrier@novell.com>

	* configure.in: version bump -> 1.2.4

2007-01-31  Mike Kestner  <mkestner@novell.com>

	* configure.in: add ilcontrast tool.
	* Makefile.am: add ilcontrast tool.

2007-01-25  Wade Berrier  <wberrier@novell.com>

	* configure.in: version bump -> 1.2.3

2006-11-17  Wade Berrier  <wberrier@novell.com>

	* configure.in: version bump -> 1.2.1

2006-10-30  Wade Berrier  <wberrier@novell.com>

	* Makefile.am: Add create-native-map to SUBDIRS so building from a
	tarball works

2006-10-24  Jonathan Pryor <jonpryor@vt.edu>


gendarme/tools/supported/gd2i/gd2i.cs
//
// gd2i - Gendarme's Defects To Ignore List
//
// Authors:
//	Sebastien Pouliot <sebastien@ximian.com>
//
// Copyright (C) 2010 Novell, Inc (http://www.novell.com)
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Text;
using System.Xml;

using NDesk.Options;

namespace Gendarme.Tools {

	class DefectsToIgnoreList {

		Dictionary<string, HashSet<string>> ignore_list = new Dictionary<string, HashSet<string>> ();

		void ReadDefects (string filename)
		{
			using (XmlTextReader reader = new XmlTextReader (filename)) {
				Dictionary<string, string> full_names = new Dictionary<string, string> ();
				// look for 'gendarme-output/rules/rule/'
				while (reader.Read () && (reader.Name != "rule"));
				do {
					full_names.Add (reader ["Name"], "R: " + reader.ReadInnerXml ());
				} while (reader.Read () && reader.HasAttributes);

				// look for 'gendarme-output/results/
				while (reader.Read () && (reader.Name != "results"));

				HashSet<string> defects = null;
				while (reader.Read ()) {
					if (!reader.HasAttributes)
						continue;

					switch (reader.Name) {
					case "rule":
						defects = new HashSet<string> ();
						ignore_list.Add (full_names [reader ["Name"]], defects);
						break;
					case "target":
						string target = reader ["Name"];
						if (target.IndexOf (' ') != -1)
							defects.Add ("M: " + target);
						else
							defects.Add ("T: " + target);
						break;
					}
				}
			}
		}

		void ReadIgnoreList (string filename)
		{
			HashSet<string> entries_for_rule = null;
			int line_no = 0;
			string rule = String.Empty;

			using (StreamReader sr = new StreamReader (filename)) {
				while (!sr.EndOfStream) {
					string line = sr.ReadLine ();
					line_no++;
					if (String.IsNullOrEmpty (line))
						continue;

					char type = line [0];
					switch (type) {
					case 'R':
						// rule
						rule = line.Trim ();
						if (!ignore_list.TryGetValue (rule, out entries_for_rule)) {
							entries_for_rule = new HashSet<string> ();
							ignore_list.Add (rule, entries_for_rule);
						}
						break;
					case '#':
						// comment - ignore
						break;
					case 'A':
					case 'T':
					case 'M':
					case 'N':
						if (rule.Length == 0) {
							if (syntax_errors_warnings) {
								Console.Error.WriteLine ("line #{0}: Entry '{1}' has no rule previously defined.",
									line_no, line);
							}
							break;
						}
						string metadata = line.Trim ();
						// metadata - remove from list
						if (!entries_for_rule.Remove (metadata) && extra_defects_warnings) {
							Console.Error.WriteLine ("line #{0}: Entry '{1}' for rule {2}' was not reported as a defect.",
								line_no, line, rule);
						}
						break;
					default:
						// unknown - report if warn is true
						if (syntax_errors_warnings) {
							Console.Error.WriteLine ("line #{0}: Invalid syntax for '{0}'.",
								line_no, line);
						}
						break;
					}
				}
			}
		}

		string BuildIgnoreData ()
		{
			if (ignore_list.Count == 0)
				return String.Empty;

			StringBuilder ignore_data = new StringBuilder ();
			foreach (KeyValuePair<string, HashSet<string>> entry in ignore_list) {
				if (entry.Value.Count == 0)
					continue;
				ignore_data.AppendLine (entry.Key);
				foreach (string metadata in entry.Value) {
					ignore_data.AppendLine (metadata);
				}
				ignore_data.AppendLine ();
			}
			return ignore_data.ToString ();
		}

		void WriteIgnoreList (string filename)
		{
			string data = BuildIgnoreData ();
			if (data.Length == 0)
				return;

			using (StreamWriter sw = File.CreateText (filename)) {
				sw.Write (data);
			}
		}

		void AppendToIgnoreList (string filename)
		{
			string data = BuildIgnoreData ();
			if (data.Length == 0)
				return;

			using (StreamWriter sw = File.AppendText (filename)) {
				sw.WriteLine ("# begin extra entries generated from defects (added on {0})", DateTime.Now);
				sw.WriteLine ();
				sw.Write (data);
				sw.WriteLine ("# end extra entries generated from defects");
			}
		}

		bool extra_defects_warnings;
		bool syntax_errors_warnings;
		bool help;
		bool version;
		bool quiet;
		string defects;
		string ignores;

		void Header ()
		{
			Assembly a = Assembly.GetExecutingAssembly ();
			Version v = a.GetName ().Version;
			if (v.ToString () != "0.0.0.0") {
				Console.WriteLine ("gd2i v{0}", v);
			} else {
				Console.WriteLine ("gd2i - Development Snapshot");
			}

			object [] attr = a.GetCustomAttributes (typeof (AssemblyCopyrightAttribute), false);
			if (attr.Length > 0)
				Console.WriteLine (((AssemblyCopyrightAttribute) attr [0]).Copyright);

			Console.WriteLine ();
		}

		static void Help ()
		{
			Console.WriteLine ("Usage: gd2i defects.xml list.ignore [--extra-defects-warnings] [--syntax-check] [--quiet] [--version] [--help]");
			Console.WriteLine ("Where");
			Console.WriteLine (" defects.xml\tThe list of defects (XML) produced by Gendarme on your project.");
			Console.WriteLine (" list.ignore\tThe file listing ignored defects entries for your project.");
			Console.WriteLine (" --extra-check\t[optional] Report ignore entries not in the defect list.");
			Console.WriteLine (" --syntax-check\t[optional] Report syntax error found in 'list.ignore' file.");
			Console.WriteLine (" --quiet\t[optional] Minimize output tro stdout.");
			Console.WriteLine (" --version\tDisplay the tool's version number.");
			Console.WriteLine (" --help\t\tShow help about the command-line options.");
			Console.WriteLine ();
		}

		byte Parse (string [] args)
		{
			var p = new OptionSet () {
				{ "syntax-check",	v => syntax_errors_warnings = v != null },
				{ "extra-check",	v => extra_defects_warnings = v != null },
				{ "quiet",		v => quiet = v != null },
				{ "version",		v => version = v != null },
				{ "h|?|help",		v => help = v != null },
			};

			List<string> files = p.Parse (args);
			if (files.Count != 2)
				return (byte) 1;

			defects = files [0];
			ignores = files [1];
			return (byte) 0;
		}

		byte Execute (string [] args)
		{
			byte result = 1;
			try {
				result = Parse (args);

				bool parsed = (result == 0);
				if (!quiet || !parsed)
					Header ();
				if (!parsed) {
					Console.WriteLine ("Invalid command-line parameters");
					Console.WriteLine ();
				}
				if (help || !parsed)
					Help ();

				if (parsed && !help && !version) {
					// read defects XML file and build an ignore list from it
					ReadDefects (defects);
					if (File.Exists (ignores)) {
						// read existing 'ignore-list' and remove them from the ignore list
						ReadIgnoreList (ignores);
						// don't touch the original (sort, comment...) but append new ignore entries
						AppendToIgnoreList (ignores);
					} else {
						WriteIgnoreList (ignores);
					}
				}
			}
			catch (Exception e) {
				Console.WriteLine ("Unhandled exception caught: {0}", e);
				result = 2;
			}
			return result;
		}

		static int Main (string [] args)
		{
			return new DefectsToIgnoreList ().Execute (args);
		}
	}
}

